fem/restriction.hpp

```diff
@@ -74,6 +74,12 @@ class ElementRestriction : public Operator
    /** Fill the J and Data arrays of SparseMatrix corresponding to the sparsity
        pattern given by this ElementRestriction, and the values of ea_data. */
    void FillJAndData(const Vector &ea_data, SparseMatrix &mat) const;
+
+   /**
+    * @brief GatherMap
+    * @return the mapping from L dofs to E dofs.
+    */
+   const Array<int> &GatherMap() const { return gatherMap; }
```

Comment:
I agree that this `ElementRestriction` has grown with interfaces not directly related to (`FillI`, `FillJAndData`, `FillSparseMatrix`). This internal array is in fact useful for all `L` and `E` kernels. I can locally derive and have access, but it does not seem the best option. Any better suggestion?

Reply:
I think it's a better option to locally, or internally to your class, or in a cpp file have a classe that derives from `ElementRestriction`. This way we don't break encapsulation, we're only responsible to maintain your code. In particular, I don't think this interface would hold if all the elements are not of the same type in the future.

I agree with you that `FillI`, `FillJAndData`, and `FillSparseMatrix` shouldn't be there (this was a bad design from me), these methods pollute this class and result in unclean code in subclasses. I would second a proposal to refactor these methods in a better design, but I don't really have a great idea how to do it at the moment. We could imagine having `FillI`, `FillJAndData`, and `FillSparseMatrix` as friend functions, but this doesn't resolve all problems, due to inheritance these functions could still be called in invalid cases (through inheritance).

Reply:
Deriving means creating at least one more `ElementRestriction`, when the idea is to re-use the `L2E_nat` or `L2E_lex` returned from `FiniteElementSpace::GetElementRestriction`.

Reply:
If you want to expose `gatherMap` go ahead, I agree that it can be useful, I'm just having doubts that this interface makes sense and isn't exposing implementation details.

Deriving solution (somewhere `private` in your class, or in `.cpp`):
```c++
class MyElementRestriction: public ElementRestriction
{
public:
   const Array<int> &GatherMap() const { return this->gatherMap; }
};
```

Friend solution:
```c++
class ElementRestriction
{
   ...
   friend myFunctionThatNeedsToAccessStuffThatYohannDoesNotWantToExpose(...);
};
```

Looking more into the details of "why" you need this, I see that you are doing kernel fusion for the `LinearFormExtension`. I thought we didn't want to do that systematically for reproducibility reasons, is there a reason why fusing kernels is okay here but not for the other kernels?
Besides, fusing kernels seems like a feature that should/could be separated from this PR. I think it would simplify a lot the review process to separate kernel fusion in a dedicated PR, while also avoiding most of the concerns about this PR. "Unfusing" the kernels would also have the advantage to have `LinearFormExtension` working for L2 spaces.

Reply:
And how can I use the `FiniteElementSpace::GetElementRestriction` if I don't want to create a redundant mapping?

Reply:
I agree that I'm using here `atomic` in the kernel - good point, we can decide what to do from this implementation.

Reply:
> And how can I use the `FiniteElementSpace::GetElementRestriction` if I don't want to create a redundant mapping?

I think you can do the following for instance:
```c++
class ElementRestriction
{
   ...
   friend Launch(...);
};
```
Again, if you are convinced that this interface makes sense go ahead, I was having doubts that this interface would make sense with different types of elements, but I guess that we will have to redesign all this anyway for mixed meshes/p-adaptivity.

Reply:
Removed, not needed anymore.